### ISSUE TITLE
if let scoping of lifetime (False Positive)

### ISSUE THREAD

**Issue by xNxExOx** [rust-lang/rust-clippy#6946]
Lint name: irrefutable_let_patterns

I tried this code:

```rust
    use glium::Display;
    let display = Display::new(builder, context, &event_loop).unwrap();
    if let gl_window = display.gl_window() {
        let window = gl_window.window();
        platform.attach_window(&mut gui_display, window, HiDpiMode::Default);
    }
    return display;
```

I expected to see this happen: **maybe** no lint?, because `if let` is used to scope lifetime of `gl_window`

Instead, this happened: Following the lint breaks the code, because lifetime in `gl_window` could be used in `Drop` of `gl_window` while `display` is still being used.

I am not even sure if this can be reasonably implemented, so it is more of an question/sugestion what you think about scoping with `if let` over manual call `Drop`, or wrapping with block like this:
```rust
    use glium::Display;
    let display = Display::new(builder, context, &event_loop).unwrap();
    {
        let gl_window = display.gl_window();
        let window = gl_window.window();
        platform.attach_window(&mut gui_display, window, HiDpiMode::Default);
    }
    return display;
```

In my **opinion** `if let` scoping is the shortest, and best explains the scope, without looking like too artificial scope.

### Meta

- `cargo clippy -V`:  clippy 0.1.52 (f5f33ec0 2021-03-19)
- `rustc -Vv`:
```
rustc 1.52.0-nightly (f5f33ec0e 2021-03-19)
binary: rustc
commit-hash: f5f33ec0e0455eefa72fc5567eb1280a4d5ee206
commit-date: 2021-03-19
host: x86_64-pc-windows-msvc
release: 1.52.0-nightly
LLVM version: 12.0.0
```

**Comment by xNxExOx** [comment]
ups Clippy just also prints compiler warnings, my mistake, I created https://github.com/rust-lang/rust/issues/83419 which should be the correct place.
